### ISSUE TITLE
The basePath is repeated twice in URL with query params

### ISSUE THREAD

**Issue by theshem** [vercel/next.js#17105]
# Bug report

## Describe the bug

The basePath is repeated twice in URL when we add query parameters. Say we have set `basePath: '/base'` in `next.config.js`, if we enter the `/base?foo=bar`, it changes to the `/base/base?foo=bar`.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Run the [https://repl.it/@theshem/NextBasePathWithQueryParams](https://repl.it/@theshem/NextBasePathWithQueryParams) 
2. Navigate to [https://nextbasepathwithqueryparams.theshem.repl.co/base?foo=bar](https://nextbasepathwithqueryparams.theshem.repl.co/base?foo=bar)

## Expected behavior

The URL should remain unchanged. I.e. `/base?foo=bar`.

## System information

- OS: macOS 10.14.5
- Browser (if applies) Chrome 85
- Version of Next.js: 9.5.3
- Version of Node.js: 12.x

**Comment by Janpot** [comment]
What should happen if you also have a page `./pages/base.js` or `./page/[dynamic].js`? Should it link to that page? Or still to the root? Should the Link behave differently depending on whether you have a page like that or not?

**Comment by theshem** [comment]
@Janpot Thank you for the response. But I'm afraid that is not the case. It is not about the `Link` component. The example above does not use it.

The issue is:  
If a `basePath` is set in `next.config.js` file, by adding query params to the URL the basePath is repeated unnecessarily. No router thing involved explicitly.

Am I missing something here?

**Comment by Janpot** [comment]
Ah, I didn't read the issue well enough it seems. Nope, this is a bug.

**Comment by jnv** [comment]
👍 for this bug. Perhaps an interesting observation here is that it seem to affect only homepage, building upon the original repl.it, I have added a subpage and the path duplication doesn't happen: https://nextbasepathwithqueryparams.janvlnas.repl.co/base/sub/file?some=query

I suspect this will be an issue with basePath matching somewhere in [next-server/lib/router/router.ts](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/lib/router/router.ts); I will give it some time to debug it today.

**Comment by jnv** [comment]
I have traced the issue down to [`prepareUrlAs`](https://github.com/vercel/next.js/blob/canary/packages/next/client/index.tsx#L116-L137) function when called via router.replace in [`componentDidMount`](https://github.com/vercel/next.js/blob/canary/packages/next/client/index.tsx#L116-L137).

When accessing the homepage at `/some/base/path?some=query`, `as` already contains the `basePath` while `url` does not:

<img width="282" alt="image" src="https://user-images.githubusercontent.com/616767/93443834-b2826f00-f8d0-11ea-91ff-8d71e76c4879.png">

But if I call the function on subpath, `as` does not contain the basePath (in this case I am accessing `/some/base/path/another/page?some=query`):

<img width="284" alt="image" src="https://user-images.githubusercontent.com/616767/93444270-d1810100-f8d0-11ea-9718-0e5bbd6092bb.png">

So I think this seems like a special treatment of `/` url in `as`. Let's see if I can do something about it.

**Comment by jnv** [comment]
So the root cause is in the `hasBasePath` function which fails to match root URL with query string:

https://github.com/vercel/next.js/blob/67b67b28b229fca0bd937d154270c6a6312e2bce/packages/next/next-server/lib/router/router.ts#L50-L52

This works for subpath when `path = '/some/base/path/another/page?some=query'` and `basePath = '/some/base/path'` since it will fulfill the `startsWith` condition.
But if `path = '/some/base/path?some=query'` it satisfies neither. I think the safest fix here would be to explicitly remove query string from `path` inside that function but perhaps some maintainer can guide me to a cleaner solution.

**Comment by jnv** [comment]
Well, since the PR with a fix didn't get any attention so far, I have implemented an ugly workaround. Just call this hook in you `index` page and it should get the work done; I lose hash part of URL if also query string is present, but other than that it seems to get the job done:

```js
import {useRouter} from 'next/router'

export function useBasepathFix(): void {
  const {basePath} = useRouter()
  if (typeof window === 'undefined') {
    return
  }
  if (window.location.pathname === `${basePath}${basePath}`) {
    window.history.replaceState(
      undefined,
      undefined,
      basePath + window.location.search + window.location.hash
    )
  }
}
```

**Comment by ijjk** [comment]
This appears to be a duplicate of https://github.com/vercel/next.js/issues/19294 which was addressed in https://github.com/vercel/next.js/pull/20596 which is available in the latest version of Next.js `v10.0.5-canary.9`, please update and give it a try!

**Comment by indotek68** [comment]
Has this been fixed? I'm not seeing it in the latest versions.

**Comment by tonydattolo** [comment]
This bug is still occurring. I'm on the latest version of NextJS. Route calls to the same page, or back to the previous page with query parameters, are doubling the url causing 404s

**Comment by tonydattolo** [comment]
> Apparently this should have been fixed in #20596. I recommend opening a new issue with repro.
> 
> Also, does the workaround mentioned in my earlier comment [#17105 (comment)](https://github.com/vercel/next.js/issues/17105#issuecomment-705003543) work for you?

Do you know where this workaround is supposed to go?

**Comment by tonydattolo** [comment]
> Apparently this should have been fixed in #20596. I recommend opening a new issue with repro.
> 
> Also, does the workaround mentioned in my earlier comment [#17105 (comment)](https://github.com/vercel/next.js/issues/17105#issuecomment-705003543) work for you?

I'll give it a shot

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
